src/main/java/co/cask/http/NettyHttpService.java

```diff
@@ -272,23 +166,108 @@ public InetSocketAddress getBindAddress() {
   @Override
   protected void shutDown() throws Exception {
     LOG.info("Stopping HTTP Service {}", serviceName);
+
     try {
-      bootstrap.shutdown();
-      if (!channelGroup.close().await(CLOSE_CHANNEL_TIMEOUT, TimeUnit.SECONDS)) {
-        LOG.warn("Timeout when closing all channels.");
-      }
+      serverChannel.close().sync();
     } finally {
+      bootstrap.config().group().shutdownGracefully();
+      bootstrap.config().childGroup().shutdownGracefully();
+      eventExecutorGroup.shutdownGracefully();
       resourceHandler.destroy(handlerContext);
-      bootstrap.releaseExternalResources();
-      if (executionHandler != null) {
-        executionHandler.releaseExternalResources();
-        ExecutorUtil.terminate(executionHandler.getExecutor());
-      }
     }
     LOG.debug("Stopped HTTP Service {} on address {}", serviceName, bindAddress);
   }
 
   /**
+   * Create {@link EventExecutorGroup} for executing handle methods.
+   *
+   * @param size size of threadPool
+   * @param keepAliveSecs  maximum time that excess idle threads will wait for new tasks before terminating
+   * @return instance of {@link EventExecutorGroup} or {@code null} if {@code size} is {@code <= 0}.
```

Comment:
I think the `@Nullable` is wrong. It doesn't seem to actually return null.

Reply:
Yeah, there are some old annotations/docs hasn't been cleanup yet.